### PR TITLE
KV Store: Graveler implement ResetKey ResetPrefix

### LINKED ISSUE #3719: KV Graveler: Implement Reset, ResetKey ResetPrefix

### PR DESCRIPTION
**Pull request by N-o-Z**
- Closes #3719

### CODE REVIEW

File: pkg/graveler/graveler.go [hunk 1]
```diff
@@ -1903,16 +1905,135 @@ func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branc
...
+				return nil, err
+			}
+		}
+
+		// replace branch tokens with new staging/sealed tokens
+		branch.StagingToken = newStagingToken
+		branch.SealedTokens = newSealedTokens
+		return branch, nil
+	})
+
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1976]: remove new line

File: pkg/graveler/graveler.go [hunk 2]
```diff
@@ -1903,16 +1905,135 @@ func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branc
...
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(branch *Branch) (*Branch, error) {
+		newSealedTokens = []StagingToken{branch.StagingToken}
+		newSealedTokens = append(newSealedTokens, branch.SealedTokens...)
+
+		// Reset the key on the new staging token
+		staged, err := g.getFromStagingArea(ctx, branch, key)
+		if err != nil {
+			if errors.Is(err, ErrNotFound) { // If key is not in staging => nothing to do
+				return nil, nil
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1960]: Is this the expected behaviour? `Reset` for a `NotFound` key in staging doesn't return an error?

File: pkg/graveler/graveler.go [hunk 3]
```diff
@@ -1903,16 +1905,135 @@ func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branc
...
+	committed, err := g.Get(ctx, repositoryID, branch.CommitID.Ref(), key)
+	if err != nil {
+		if !errors.Is(err, ErrNotFound) {
+			return err
+		}
+		isCommitted = false
+	}
+
+	if isCommitted { // entry committed and changed in staging area => override with entry from commit
+		if reflect.DeepEqual(committed, stagedValue) {
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1926]: This has really bad performance. There's an `Identity` property exactly for that.. I think we already have a `Compare`/`Equal` method for that.

File: pkg/graveler/graveler.go [hunk 4]
```diff
@@ -1903,16 +1905,135 @@ func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branc
...
+			return nil // No change
+		}
+		if err = g.StagingManager.Set(ctx, st, key, committed, true); err != nil {
+			return err
+		}
+	} else if !isCommitted && stagedValue != nil { // entry not committed and changed in staging area => override with tombstone
+		if err = g.StagingManager.Set(ctx, st, key, nil, true); err != nil {
+			return err
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1934]:
```suggestion
		return g.StagingManager.Set(ctx, st, key, nil, true)
```

File: pkg/graveler/graveler.go [hunk 5]
```diff
@@ -1903,16 +1905,135 @@ func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branc
...
+		isCommitted = false
+	}
+
+	if isCommitted { // entry committed and changed in staging area => override with entry from commit
+		if reflect.DeepEqual(committed, stagedValue) {
+			return nil // No change
+		}
+		if err = g.StagingManager.Set(ctx, st, key, committed, true); err != nil {
+			return err
+		}
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1931]:
```suggestion
		return g.StagingManager.Set(ctx, st, key, committed, true); err != nil
```

File: pkg/graveler/graveler.go [hunk 6]
```diff
@@ -1903,16 +1905,135 @@ func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branc
...
+	}
+	if isProtected {
+		return ErrWriteToProtectedBranch
+	}
+	// New sealed tokens list after change includes current staging token
+	newSealedTokens := make([]StagingToken, 0)
+	newStagingToken := generateStagingToken(repositoryID, branchID)
+
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(branch *Branch) (*Branch, error) {
+		newSealedTokens = []StagingToken{branch.StagingToken}
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1953]: Let's add a compaction task, this implementation makes it necessary.

File: pkg/graveler/graveler.go [hunk 7]
```diff
@@ -1903,16 +1905,135 @@ func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branc
...
+		var wg multierror.Group
+		for itr.Next() {
+			value := itr.Value()
+			if !bytes.HasPrefix(value.Key, key) { // We passed the prefix - exit the loop
+				break
+			}
+			wg.Go(func() error {
+				err = g.resetKey(ctx, repositoryID, branch, value.Key, value.Value, newStagingToken)
+				if err != nil {
+					if !errors.Is(err, ErrNotFound) { // Not found in staging => ignore
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:2015]: `g.resetKey` can't return `ErrNotFound`

File: pkg/graveler/graveler.go [hunk 8]
```diff
@@ -1903,16 +1905,135 @@ func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branc
...
+		if err != nil {
+			return nil, err
+		}
+
+		// replace branch tokens with new staging/sealed tokens
+		branch.StagingToken = newStagingToken
+		branch.SealedTokens = newSealedTokens
+		return branch, nil
+	})
+
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:2032]: drop line

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Great PR, it leverages many of the constructs we made for the new Graveler implementation. I think it's very important to test this delicate functionalities properly (can be in another PR though)

File: pkg/graveler/graveler.go (hunk 6 above)
Comment (N-o-Z) [reply pkg/graveler/graveler.go:1953]: https://github.com/treeverse/lakeFS/issues/3780

File: pkg/graveler/graveler.go (hunk 2 above)
Comment (N-o-Z) [reply pkg/graveler/graveler.go:1960]: Yes - but DropKey should also not return an error - fixed

File: pkg/graveler/graveler.go [hunk 9]
```diff
@@ -1903,16 +1904,129 @@ func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branc
...
+		if bytes.Equal(committed.Identity, stagedValue.Identity) {
+			return nil // No change
+		}
+		return g.StagingManager.Set(ctx, st, key, committed, true)
+		// entry not committed and changed in staging area => override with tombstone
+		// If not committed and staging == tombstone => ignore
+	} else if !isCommitted && stagedValue != nil {
+		return g.StagingManager.Set(ctx, st, key, nil, true)
+	}
 	return nil
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1934]:
```suggestion
	} else if !isCommitted && stagedValue != nil {
		// entry not committed and changed in staging area => override with tombstone
		return g.StagingManager.Set(ctx, st, key, nil, true)
	}

	// If not committed and staging == tombstone => ignore
	return nil
```

File: pkg/graveler/graveler.go (hunk 2 above)
Comment (itaiad200) [reply pkg/graveler/graveler.go:1957]: Cool - I don't think we ever call it from KVGraveler

**Review by itaiad200 — APPROVED** [review]
It's coming home

**Pull request merged**